python/test/unit/fem/test_petsc_nonlinear_assembler.py

```diff
@@ -718,33 +537,30 @@ def monolithic():
                 dirichletbc(u_bc_1, bdofsW0_P2_1, W.sub(0)),
             ]
 
-            Jmat = create_matrix(J)
-            Pmat = create_matrix(P)
-            Fvec = create_vector(F)
-
-            snes = PETSc.SNES().create(MPI.COMM_WORLD)
-            snes.setTolerances(rtol=1.0e-15, max_it=20)
-            snes.getKSP().setType("minres")
-
-            problem = NonlinearPDE_SNESProblem(F, J, U, bcs, P=P)
-            snes.setFunction(problem.F_mono, Fvec)
-            snes.setJacobian(problem.J_mono, J=Jmat, P=Pmat)
-
             U.sub(0).interpolate(initial_guess_u)
             U.sub(1).interpolate(initial_guess_p)
 
-            x = create_vector(F)
-            x.array[:] = U.x.petsc_vec.array_r
-
-            snes.solve(None, x)
-            assert snes.getConvergedReason() > 0
-            snes.destroy()
-            Jnorm = Jmat.norm()
-            Fnorm = Fvec.norm()
+            snes_options = {
+                "snes_rtol": 1.0e-15,
+                "snes_max_it": 20,
+                "ksp_type": "minres",
+                "snes_monitor": None,
+            }
+            solver = dolfinx.nls.petsc.SNESSolver(
+                F,
+                U,
+                J=J,
+                bcs=bcs,
+                P=P,
+                assembly_type=dolfinx.fem.petsc.AssemblyType.standard,
+                snes_options=snes_options,
+            )
+            x, converged_reason, _ = solver.solve()
+            assert converged_reason > 0
+            solver.copy_vec_to_function(x, U)
```

Comment:
If we can agree that `SNESSolver` should be `NonlinearProblem` then this call should go inside solve and also solve should do any necessary ghost updates, in the same way that `LinearProblem` works.

Reply:
Let's do that in this PR so that we can review this important set of changes as a whole.